fx quant: support for ndim followed by binary op

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53875 fx quant: remove unnecessary dequants for getattr
* #53614 fx quant: clean up names of quantize handlers
* #53196 fx quant: ensure fp32 sum works correctly with quantized layers
* #53187 fx quant: fix using size of quant layer in torch._assert
* **#53120 fx quant: support for ndim followed by binary op**

Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

The fix implemented in this PR is adding a heuristic on whether a node has all of its args
known to be not tensors.  In the longer term, this may need to be improved to adding reasoning
about the return type of various ops, propagating it through the graph, and updating the rules to use it.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

vkuzo committed:
[not for land yet] hacky fix for x.ndim followed by sub

Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

For now, we hack a bandaid to unblock some teams, none of this is for
land.  We will have to think of a better fix which is landable (TBD).

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land yet] hacky fix for x.ndim followed by sub"


Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

For now, we hack a bandaid to unblock some teams, none of this is for
land.  We will have to think of a better fix which is landable (TBD).

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -192,9 +192,23 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
+                # add_args = (*load_arg(quantized=True)(self.binary_op_node.args), scale_arg, zero_point_arg)
+                # op = quantizer.quantized_graph.create_node(
+                    # 'call_function', self.quantized_binary_op, add_args, kwargs)
+                if True:
```
We should figure out the best implementation here.  A challenge with the current approach is that there is a new type of arg, "potentially quantized".  So, we either need to increase the # of types of args from two to three, or think of a different approach to this PR.

jerryzh168 replied:
yeah this part looks hacky, is it possible to identify that a node is produced by ndim, size etc. and filter out these nodes during loading? something like:

```
int_node_idxs = [1]
all_arg_idxs = [1, 2]
quantized_idxs = [2]
... = load_arg(quantized=quantized_idxs)...
```

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -318,6 +318,60 @@ def node_arg_is_bias(node: Node, arg: Any) -> bool:
                     return True
     return False
 
+def all_node_args_have_no_tensors(node: Node) -> bool:
```
in the short term, this sounds OK.  In the longer term, we might need to evaluate introducing an output type for nodes, which is known for some nodes (and maybe not others) and can propagate through various ops.  For example, we know that the output of `x.size(...)` is always an integer, etc.

jerryzh168 reviewed (approved):
The try except part looks hacky, but the rest seems to be OK I think.

vkuzo committed:
Update on "[not for land yet] unbreak fx quant for ndim followed by binary op"


Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

The short term fix implemented in this PR is adding a heuristic on whether a node has all of its args
known to be not tensors.  In the longer term, this may need to be improved to adding reasoning
about the return type of various ops, propagating it through the graph, and updating the rules to use it.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land yet] unbreak fx quant for ndim followed by binary op"


Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

The short term fix implemented in this PR is adding a heuristic on whether a node has all of its args
known to be not tensors.  In the longer term, this may need to be improved to adding reasoning
about the return type of various ops, propagating it through the graph, and updating the rules to use it.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/utils.py:
```diff
@@ -330,3 +330,46 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
+                    found_one_tensor = found_one_tensor or \
+                        (not this_list_el_args_have_no_tensors)
+        elif isinstance(arg, int):
+            pass
```
pass? should we remove?

jerryzh168 commented on torch/quantization/fx/utils.py:
```diff
@@ -330,3 +330,46 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
+                this_arg_args_have_no_tensors = all_node_args_have_no_tensors(arg)
+                found_one_tensor = found_one_tensor or \
+                    (not this_arg_args_have_no_tensors)
+            else:
```
why is this True?

vkuzo replied:
yeah, this is a heuristic, we are assuming all "unknown" types are tensors

btw, in the future, we can delete this code and switch to https://github.com/pytorch/pytorch/pull/53653, once it is ready

vkuzo committed:
Update on "fx quant: support for ndim followed by binary op"


Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

The fix implemented in this PR is adding a heuristic on whether a node has all of its args
known to be not tensors.  In the longer term, this may need to be improved to adding reasoning
about the return type of various ops, propagating it through the graph, and updating the rules to use it.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

[ghstack-poisoned]
